Fallback to fetching public key from certificate not always applied?

(moving this from a comment on a closed issue to a new thread)
Request: Make the fallback to fetching public key from certificate work with OpenSSL's SSL_CTX_use_Privatekey().

This fallback was apparently implemented by #129 but that does not seem to work when an EC private key is passed to `SSL_CTX_use_Privatekey()` in OpenSSL even if matching certificate is present with same ID. In this case `p11prov_ec_match()` gets called which fails to find the public key via `find_associated_obj()`. Would be nice to get the fallback applied here too for wider usability. 

Whereas, when signing using "openssl pkeyutl" the fallback does work and appears to follow a different code path.

Tested using git master branch and a soft token (softhsm2) where only private key (EC) and certificate were imported into the token using `p11tool` -- public key was not explicitly imported. The same token works with `pkcs11-helper` library. 

Unlike `p11tool`, `softhsm2-util` imports keypair in one step and creates both public and private key objects in the token --- in that case there are no such errors, obviously.

What does not work is setting an EC certificate and key in an SSL context using OpenSSL API in C. In short this involves two steps (see below for a full test program):
(i) fetch a certificate and private key using OpenSSL STORE API and a `pkcs11:` URI with `pkcs11-provider` loaded. 
(ii) set the certificate and the key in an SSL context. The latter (i.e., setting the key) fails when public key is not on the token, succeeds otherwise.

[Here](https://gist.github.com/selvanair/ebe0263cd52b501be2031aa316d5cf7d) is a sample program that demonstrates this. I tested it using a soft token (softhsm2) which contains a single key and cert imported with the same id = 01 so that the URI used is hard-coded as simply: "pkcs11:id=01".

OpenSSL error:
```
4037945B88740000:error:40800000:pkcs11:find_associated_obj:pkcs11:../src/objects.c:1136:Error in C_FindObjects (count=0)
4037945B88740000:error:40800005:pkcs11:match_public_keys:General Error:../src/objects.c:2314:Could not find associated public key object
4037945B88740000:error:05800074:x509 certificate routines:X509_check_private_key:key values mismatch:../crypto/x509/x509_cmp.c:405:
```

pkcs11-provider debug output just before the above error:
```
[../src/keymgmt.c:1343] p11prov_ec_has(): ec has 0x5f1ebfae75c0 2
[../src/keymgmt.c:1254] p11prov_ec_new(): ec new
[../src/keymgmt.c:1377] p11prov_ec_import(): ec import 0x5f1ebfaf1980
[../src/keymgmt.c:1365] p11prov_ec_match(): ec match 0x5f1ebfaf1980 0x5f1ebfae75c0 6
[../src/objects.c:1096] find_associated_obj(): Find associated object
[../src/session.c:867] p11prov_get_session(): Get session on slot 1844729915, reqlogin=true, rw=false
[../src/session.c:876] p11prov_get_session(): single-shot request for slot 1844729915
[../src/session.c:596] check_slot(): Checking Slot id=1844729915, uri=(nil), mechtype=ffffffffffffffff, rw=false)
[../src/provider.c:595] p11prov_ctx_login_behavior(): login_behavior = 0
[../src/interface.gen.c:305] p11prov_GetSessionInfo(): Calling C_GetSessionInfo
[../src/interface.gen.c:305] p11prov_GetSessionInfo(): Calling C_GetSessionInfo
[../src/interface.gen.c:533] p11prov_FindObjectsInit(): Calling C_FindObjectsInit
[../src/interface.gen.c:559] p11prov_FindObjects(): Calling C_FindObjects
[../src/interface.gen.c:583] p11prov_FindObjectsFinal(): Calling C_FindObjectsFinal
[../src/objects.c:1136] find_associated_obj(): Error: 0x00000000; Error in C_FindObjects (count=0)
[../src/objects.c:2314] match_public_keys(): Error: 0x00000005; Could not find associated public key object
```

So it appears `p11prov_ec_match()` fails in `find_associated_obj()` called from `match_public_keys()`.

--- a/src/objects.c
+++ b/src/objects.c
@@ -2279,6 +2279,81 @@ static int cmp_public_key_values(P11PROV_OBJ *pub_key1, P11PROV_OBJ *pub_key2)
     return ret;
 }
 
+static int match_key_with_cert(P11PROV_OBJ *priv_key, P11PROV_OBJ *pub_key)
+{
+    P11PROV_OBJ *cert;
+    CK_ATTRIBUTE attrs[2] = { 0 };
+    CK_ATTRIBUTE *x;
+    int num = 0;
+    int ret = RET_OSSL_ERR;
+
+    cert = find_associated_obj(priv_key->ctx, priv_key, CKO_CERTIFICATE);
+    if (!cert) {
+        P11PROV_raise(priv_key->ctx, CKR_GENERAL_ERROR,
+                      "Could not find associated certificate object");
+        return RET_OSSL_ERR;
+    }
+
+    switch (pub_key->data.key.type) {
+    case CKK_RSA:
+        attrs[0].type = CKA_MODULUS;
+        attrs[1].type = CKA_PUBLIC_EXPONENT;
+        num = 2;
+        break;
+    case CKK_EC:
+    case CKK_EC_EDWARDS:
+        attrs[0].type = CKA_P11PROV_PUB_KEY;
+        num = 1;
+        break;
+    }
+
+    ret = get_all_from_cert(cert, attrs, num);
+    if (ret != CKR_OK) {
+        P11PROV_raise(priv_key->ctx, ret,
+                      "Failed to get public attrs from cert");
+        ret = RET_OSSL_ERR;
+        goto done;
+    }
+
+    switch (pub_key->data.key.type) {
+    case CKK_RSA:
+        x = p11prov_obj_get_attr(pub_key, CKA_MODULUS);
+        if (!x || x->ulValueLen != attrs[0].ulValueLen
+            || memcmp(x->pValue, attrs[0].pValue, x->ulValueLen) != 0) {
+            ret = RET_OSSL_ERR;
+            goto done;
+        }
+
+        x = p11prov_obj_get_attr(pub_key, CKA_PUBLIC_EXPONENT);
+        if (!x || x->ulValueLen != attrs[1].ulValueLen
+            || memcmp(x->pValue, attrs[1].pValue, x->ulValueLen) != 0) {
+            ret = RET_OSSL_ERR;
+            goto done;
+        }
+
+        ret = RET_OSSL_OK;
+        break;
+    case CKK_EC:
+    case CKK_EC_EDWARDS:
+        x = p11prov_obj_get_attr(pub_key, CKA_P11PROV_PUB_KEY);
+        if (!x || x->ulValueLen != attrs[0].ulValueLen
+            || memcmp(x->pValue, attrs[0].pValue, x->ulValueLen) != 0) {
+            ret = RET_OSSL_ERR;
+            goto done;
+        }
+
+        ret = RET_OSSL_OK;
+        break;
+    }
+
+done:
+    for (int i = 0; i < num; i++) {
+        OPENSSL_free(attrs[i].pValue);
+    }
+    p11prov_obj_free(cert);
+    return ret;
+}
+
 static int match_public_keys(P11PROV_OBJ *key1, P11PROV_OBJ *key2)
 {
     P11PROV_OBJ *pub_key, *assoc_pub_key;
@@ -2313,7 +2388,10 @@ static int match_public_keys(P11PROV_OBJ *key1, P11PROV_OBJ *key2)
     if (!assoc_pub_key) {
         P11PROV_raise(priv_key->ctx, CKR_GENERAL_ERROR,
                       "Could not find associated public key object");
-        return RET_OSSL_ERR;
+
+        /* some tokens only store the public key in a cert and not in a
+         * separate public key object */
+        return match_key_with_cert(priv_key, pub_key);
     }
 
     if (assoc_pub_key->data.key.type != pub_key->data.key.type) {

--- a/tests/meson.build
+++ b/tests/meson.build
@@ -104,6 +104,7 @@ test_programs = {
   'tsession': ['tsession.c'],
   'tgenkey': ['tgenkey.c'],
   'tlsctx': ['tlsctx.c', 'util.c'],
+  'tlssetkey': ['tlssetkey.c', 'util.c'],
   'tdigests': ['tdigests.c'],
   'treadkeys': ['treadkeys.c'],
   'tcmpkeys': ['tcmpkeys.c', 'util.c'],

--- a/tests/tlssetkey.c
+++ b/tests/tlssetkey.c
@@ -1,0 +1,53 @@
+/* Copyright (C) 2024 Simo Sorce <simo@redhat.com>
+   SPDX-License-Identifier: Apache-2.0 */
+
+#include <stdio.h>
+#include <stdbool.h>
+#include <openssl/ssl.h>
+#include <openssl/evp.h>
+#include <openssl/rsa.h>
+#include <openssl/core_names.h>
+#include "util.h"
+
+int main(int argc, char *argv[])
+{
+    EVP_PKEY *pkey = NULL;
+    X509 *cert = NULL;
+    SSL_CTX *ctx;
+    int ret = 0;
+
+    if (argc != 3) {
+        fprintf(stderr, "Usage: tlssetkey [certuri] [pkeyuri]\n");
+        exit(EXIT_FAILURE);
+    }
+    cert = load_cert(argv[1]);
+    pkey = load_key(argv[2]);
+
+    ctx = SSL_CTX_new(TLS_client_method());
+    if (!ctx) {
+        fprintf(stderr, "Failed to create SSL Context\n");
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+
+    ret = SSL_CTX_use_certificate(ctx, cert);
+    if (ret != 1) {
+        fprintf(stderr, "Failed to set Certificate");
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+
+    ret = SSL_CTX_use_PrivateKey(ctx, pkey);
+    if (ret != 1) {
+        fprintf(stderr, "Failed to set Private Key");
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+
+    fprintf(stderr, "Cert and Key successfully set on TLS Context!\n");
+
+    SSL_CTX_free(ctx);
+    EVP_PKEY_free(pkey);
+    X509_free(cert);
+    exit(EXIT_SUCCESS);
+}

--- a/tests/ttls
+++ b/tests/ttls
@@ -7,6 +7,14 @@ source "${TESTSSRCDIR}/helpers.sh"
 title PARA "Test SSL_CTX creation"
 $CHECKER "${TESTBLDDIR}/tlsctx"
 
+title PARA "Test setting cert/keys on TLS Context"
+$CHECKER "${TESTBLDDIR}/tlssetkey" "${ECCRTURI}" "${ECPRIURI}"
+
+if [ -n "$ECBASE2URI" ]; then
+    title PARA "Test setting cert/keys on TLS Context w/o pub key"
+    $CHECKER "${TESTBLDDIR}/tlssetkey" "${ECCRT2URI}" "${ECPRI2URI}"
+fi
+
 title PARA "Test an actual TLS connection"
 
 rm -f "${TMPPDIR}/s_server_output"

--- a/tests/util.c
+++ b/tests/util.c
@@ -92,3 +92,49 @@ EVP_PKEY *load_key(const char *uri)
 
     return key;
 }
+
+X509 *load_cert(const char *uri)
+{
+    OSSL_STORE_CTX *store;
+    OSSL_STORE_INFO *info;
+    X509 *cert = NULL;
+
+    if (!uri) {
+        fprintf(stderr, "Invalid NULL uri");
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+
+    store = OSSL_STORE_open(uri, NULL, NULL, NULL, NULL);
+    if (store == NULL) {
+        fprintf(stderr, "Failed to open store: %s\n", uri);
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+
+    for (info = OSSL_STORE_load(store); info != NULL;
+         info = OSSL_STORE_load(store)) {
+        int type = OSSL_STORE_INFO_get_type(info);
+
+        if (cert != NULL) {
+            fprintf(stderr, "Multiple certs matching URI: %s\n", uri);
+            exit(EXIT_FAILURE);
+        }
+
+        switch (type) {
+        case OSSL_STORE_INFO_CERT:
+            cert = OSSL_STORE_INFO_get1_CERT(info);
+            break;
+        }
+        OSSL_STORE_INFO_free(info);
+    }
+
+    if (cert == NULL) {
+        fprintf(stderr, "Failed to load cert from URI: %s\n", uri);
+        ossl_err_print();
+        exit(EXIT_FAILURE);
+    }
+    OSSL_STORE_close(store);
+
+    return cert;
+}

--- a/tests/util.h
+++ b/tests/util.h
@@ -3,3 +3,4 @@
 
 void ossl_err_print(void);
 EVP_PKEY *load_key(const char *uri);
+X509 *load_cert(const char *uri);

